desk/lib/discipline.hoon

```diff
@@ -0,0 +1,307 @@
+  ?:  =(%$ i.a)  |
+  ?:  =(%$ i.b)  &
+  (aor i.a i.b)
+=.  facts   (sort facts por)
+=.  scries  (sort scries por)
+::
+=/  mark-map=(map mark [strict=? =type])
+  (~(gas by *(map mark [? type])) marks)
+=/  mark-types=(map mark type)
+  (~(run by mark-map) tail)
+::
+|=  inner=agent:gall
+|^  agent
+::
++$  card  card:agent:gall
+::  +match-path: is .test a template-prefix of .real?
+::
+++  match-path
+  |=  [test=path real=path]
+  ^-  ?
+  ?|  ?=(~ test)
+  ?&  ?=(^ real)
+      |(=(%$ i.test) =(i.test i.real))
+      $(test t.test, real t.real)
+  ==  ==
+::  +expected-fact-marks: marks from the first path in .facts matching .path
+::
+++  expected-fact-marks
+  |=  =path
+  ^-  (unit (list mark))
+  ?~  facts  ~
+  ?:  (match-path path.i.facts path)
+    `marks.i.facts
+  $(facts t.facts)
+::
+::  +expected-scry-mark: marks from the first path in .scries matching .path
+::
+++  expected-scry-mark
+  |=  =path
+  ^-  (unit mark)
+  ?~  scries  ~
+  ?:  (match-path path.i.scries path)
+    `mark.i.scries
+  $(scries t.scries)
+::
+++  helper
+  |_  [=bowl:gall marks=(map mark type)]
+  ++  log  ~(. logs our.bowl /~/discipline/logs)
+  ::  +inner-bowl: bowl without /~/discipline subs
+  ::
+  ++  inner-bowl
+    %_  bowl
+        wex
+      %-  ~(gas by *boat:gall)
+      %+  skip  ~(tap by wex.bowl)
+      |=  [[=wire *] *]
+      ?=([%~.~ %discipline *] wire)
+    ==
+  ::  +check-cards: check facts among cards
+  ::
+  ++  check-cards
+    |=  [watch-path=(unit path) cards=(list card)]
+    ^+  cards
+    (zing (turn cards (cury check-card watch-path)))
+  ::  +check-card: check facts for mismatching marks, log if so
+  ::
+  ++  check-card
+    |=  [watch-path=(unit path) =card]
+    ^-  (list ^card)
+    :-  card
+    ?.  ?=([%give %fact *] card)  ~
+    =?  paths.p.card  ?=(~ paths.p.card)
+      ~?  >>  ?=(~ watch-path)  [%discipline dap=dap.bowl %initial-fact-but-no-watch-path]
+      (drop watch-path)
+    %+  murn  paths.p.card
+    |=  =path
+    ^-  (unit ^card)
+    ?~  ex-marks=(expected-fact-marks path)
+      ~&  >>  [%discipline dap=dap.bowl %unknown-path path=path]
+      `(tell:log %warn ~['discipline: unknown path' >path<] ~)
+    ?.  ?|  =(~ u.ex-marks)  ::  empty list allows any mark
+            ?=(^ (find ~[p.cage.p.card] u.ex-marks))
+        ==
+      =+  n=[path=path mark=p.cage.p.card allowed=u.ex-marks]
+      ~&  >>>  [%discipline dap=dap.bowl %bad-fact-mark n]
+      `(tell:log %crit ~['discipline: bad fact mark' >n<] ~)
+    ?~  type=(~(get by marks) p.cage.p.card)
+      ::NOTE  if we don't have a type for this mark,
+      ::      ignore it, since we can't test it
+      ~
+    ?:  (~(nest ut u.type) | p.q.cage.p.card)
+      ~
+    =+  n=[on-path=path with-mark=p.cage.p.card %nest-fail]
+    ~&  >>>  [%discipline dap=dap.bowl %bad-fact-data n]
+    `(tell:log %crit ~['discipline: bad fact data' >n<] ~)
+  ::  +check-scry: validate result against known marks
+  ::
+  ++  check-scry
```

Comment:
Do we not intend to log mismatches in this arm as well?

Reply:
Handling scries is read-only, so we cannot possibly log anything.